core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -1025,9 +1026,13 @@ class KafkaServer(
    */
   private def checkpointBrokerMetadata(brokerMetadata: ZkMetaProperties) = {
     for (logDir <- config.logDirs if logManager.isLogDirOnline(new File(logDir).getAbsolutePath)) {
+      val props = brokerMetadata.toProperties.clone().asInstanceOf[Properties]
+      // If the Uuid is not set then we set it here because originally we ignored
+      // directories with no meta.properties file and we are creating it here.
+      props.setProperty("directory.id", logManager.directoryId(logDir).getOrElse(Uuid.randomUuid()).toString)
```

Comment:
LogManager is already generating a directory ID if missing, in `directoryIds`.

Reply:
Okay, I fixed LogManager to generate the directory ID for all cases where it is missing. This includes if the meta.properties is missing which can only occur if we are in ZK mde. It does not create the missing meta.properties file until after the broker is registered with ZK.